pkg/profiler/profiler.go

```diff
@@ -405,33 +373,77 @@ func (p *CgroupProfiler) profileLoop(ctx context.Context, captureTime time.Time)
 		kernelLocations = []*profile.Location{}
 		userLocations   = map[uint32][]*profile.Location{} // PID -> []*profile.Location
 		locationIndices = map[[2]uint64]int{}              // [PID, Address] -> index in locations
+
+		// Variables needed for eBPF map batch iteration.
+		countKeysPtr = unsafe.Pointer(&p.countKeys[0])
+		nextCountKey = uintptr(1)
 	)
 
-	// TODO(kakkoyun): Use libbpf batch functions.
+	// Reset count keys before collecting new traces from the kernel.
+	memsetCountKeys(p.countKeys, stackCountKey{})
+
+	batchSize := 0
 	it := p.bpfMaps.counts.Iterator()
 	for it.Next() {
-		// This byte slice is only valid for this iteration, so it must be
-		// copied if we want to do anything with it outside this loop.
-		keyBytes := it.Key()
-
-		var key stackCountKey
-		// NOTICE: This works because the key struct in Go and the key struct in C has exactly the same memory layout.
-		// See the comment in stackCountKey for more details.
-		if err := binary.Read(bytes.NewBuffer(keyBytes), p.byteOrder, &key); err != nil {
-			return fmt.Errorf("read stack count key: %w", err)
+		batchSize++
+	}
+	if err := it.Err(); err != nil {
+		return fmt.Errorf("iterate over counts map: %w", err)
+	}
+
+	if batchSize == 0 {
+		return nil
+	}
+	level.Debug(p.logger).Log("msg", "fetching stack trace counts in batch", "batchSize", batchSize)
+	time.Sleep(1 * time.Second)
```

Comment:
ah sorry I see that https://github.com/parca-dev/parca-agent/pull/395 does exactly this

Reply:
Yes :). BTW, obvious perhaps now, but forgot to mention that running `parca-agent` under a debugger or adding this sleep here "made things work", because it would allow for the requested count to be less than the actual amount of elements in the map, making it work with the previous bogus error checking.

I think that the counting logic that we previously had was wrong (possibly because of the same error handling problems we spotted in libbpfgo) which made us overcount. Did not manage to prove this, but I am pretty sure this is the reason!

Reply:
It will be gone when we merge https://github.com/parca-dev/parca-agent/pull/395/files onto this